docs/syntax.md

```diff
@@ -448,6 +450,16 @@ class Ref[A] < Object
 end
 ```
 
+For classes with type parameters, you may specify if they are "invariant" (default), "covariant" (`out`) or "contravariant" (`in`). See [this definition of covariance and contravariance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)).
+
+For example, an `Array` of `String` can almost be considered to be an `Array` of `Object`, but not the reverse, so we have:
+
+```
+class Array[out T]
```

Comment:
I'm not very sure if `Array[T]` is a good example here because it is `unchecked`... How about using `Range[T]` because it is truly covariant?

Reply:
Could you explain the purpose of `unchecked`?
Why is `Array/Hash` unchecked but `Range` is not?
I could be introduced that notion at the same time in the doc.

Reply:
`Array` and `Hash` are mutable and thus making the type parameters covariant breaks the type system.

```rb
a = [1,2,3]       # Assume type of the variable `a` is Array[Integer]
b = a             # Assume type of the variable `b` is Array[Integer | String]

b.push("foo")     # This is okay because b: Array[Integer | String]

a.last + 1       # 💥 
```

However, they are treated as covariant in popular languages (example: Java and TypeScript) because it's more useful and keeping them invariant confuses users.

`unchecked` is to skip the variance validation on type parameters.